src/Enzyme.jl

```diff
@@ -925,18 +926,112 @@ grad = gradient(Reverse, only ∘ f, (a = 2.0, b = [3.0], c = "str"))
 
 # output
 
-(a = 3.0, b = [2.0], c = "str")
+(a = 3.0, b = [2.0], c = nothing)
 ```
 """
 @inline function gradient(::ReverseMode, f::F, x::X) where {F, X}
     if Compiler.active_reg_inner(X, #=seen=#(), #=world=#nothing, #=justActive=#Val(true)) == Compiler.ActiveState
         dx = Ref(make_zero(x))
         autodiff(Reverse, f∘only, Active, Duplicated(Ref(x), dx))
-        return only(dx)
+        return strip_types(only(dx))
     else
         dx = make_zero(x)
         autodiff(Reverse, f, Active, Duplicated(x, dx))
-        return dx
+        return strip_types(dx)
+    end
+end
+
+
+"""
+    strip_types(x)
+
+Aims to replace all structs with plain NamedTuples,
+except for types which are their own natural cotangent representations,
+such as Complex, Diagonal, Transpose.
+
+Maps non-diff elements to nothing (whereas `make_zero` preserves their values).
+
+Ignores mutability (whereas `make_zero` uses a cache to preserve identifications).
+
+!!! warning
+    This function is used by `gradient` to clean up its return, but it is not public at present.
+    Extending it in your package is not recommended.
+
+# Examples
+```
+julia> [1,2,3.]' |> strip_types
+1×3 adjoint(::Vector{Float64}) with eltype Float64:
+ 1.0  2.0  3.0
+
+julia> [1,2,3]' |> strip_types
+(parent = nothing,)
+
+julia> Symmetric([1 2; 3 4.]) |> strip_types
+(data = [1.0 2.0; 3.0 4.0], uplo = nothing)
+
+julia> LinRange(1,2,3) |> strip_types
+(start = 1.0, stop = 2.0, len = nothing, lendiv = nothing)
+
+julia> Ref([[1,2,3.]', (4:6)', (7.0, 16//2)]) |> strip_types
+(x = Any[[1.0 2.0 3.0], (parent = (start = nothing, stop = nothing),), (7.0, (num = nothing, den = nothing))],)
+```
+"""
+function strip_types end
+
+export strip_types
+
+const GoodNum = Union{AbstractFloat, Complex{<:AbstractFloat}}
+strip_types(x::GoodNum) = x
+strip_types(x::Array{<:GoodNum}) = x
+# strip_types(x::DenseArray{<:GoodNum}) = x
+# strip_types(x::Base.TwicePrecision{<:GoodNum}) = x
+
+# Non-differentiable types
+strip_types(x::Integer) = nothing # zero(x)
+strip_types(x::Union{Symbol, Char, AbstractString, Nothing}) = nothing # x
+strip_types(x::Array{<:Integer}) = nothing
+strip_types(x::Array{<:Union{Symbol, Char, AbstractString, Nothing}}) = nothing
+
+# Containers to recurse into
+strip_types(x::Union{Tuple, NamedTuple, Array}) = map(strip_types, x)  # need to worry about undef?
```

Comment:
This won't work, for example, for a recursive type (and will instead infinite loop presumably?)